spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/LabeledSourceData.java

```diff
@@ -50,43 +58,49 @@ public final SourceData compute(Map<String, String> labels, ConfigUtils.Prefix p
 			}
 			data = dataSupplier(labels, profiles);
 
-			// need this check because when there is no data, the name of the property
-			// source is using provided labels,
-			// unlike when the data is present: when we use secret names
-			if (data.names().isEmpty()) {
-				String names = labels.keySet()
-					.stream()
-					.sorted()
-					.collect(Collectors.joining(PROPERTY_SOURCE_NAME_SEPARATOR));
-				return SourceData.emptyRecord(ConfigUtils.sourceName(target, names, namespace));
+			LinkedHashSet<String> sourceNames = data.names();
+			Map<String, Object> sourceDataForSourceName = data.data();
+			sourceDataName = sourceDataName(target, sourceNames, namespace);
+
+			if (sourceNames.isEmpty()) {
+				return emptySourceData(labels, target, namespace);
+			}
+
+			if (prefix.getName().equals(Prefix.DEFAULT.getName())) {
+				return new SourceData(sourceDataName, defaultFlattenedSourceData(sourceNames, sourceDataForSourceName));
+			}
+
+			if (prefix.getName().equals(Prefix.KNOWN.getName())) {
+				return new SourceData(sourceDataName,
+						prefixFlattenedSourceData(sourceNames, sourceDataForSourceName, prefix.prefixProvider().get()));
 			}
 
-			if (prefix != ConfigUtils.Prefix.DEFAULT) {
-
-				String prefixToUse;
-				if (prefix == ConfigUtils.Prefix.KNOWN) {
-					prefixToUse = prefix.prefixProvider().get();
-				}
-				else {
-					prefixToUse = data.names()
-						.stream()
-						.sorted()
-						.collect(Collectors.joining(PROPERTY_SOURCE_NAME_SEPARATOR));
-				}
-
-				PrefixContext prefixContext = new PrefixContext(data.data(), prefixToUse, namespace, data.names());
-				return ConfigUtils.withPrefix(target, prefixContext);
+			if (prefix.getName().equals(Prefix.DELAYED.getName())) {
+				return new SourceData(sourceDataName, nameFlattenedSourceData(sourceNames, sourceDataForSourceName));
 			}
+
+			throw new IllegalArgumentException("Unsupported prefix: " + prefix);
 		}
 		catch (Exception e) {
 			LOG.warn("Failure in reading labeled sources");
 			onException(failFast, e);
-			data = new MultipleSourcesContainer(data.names(), Map.of(ERROR_PROPERTY, "true"));
+			return new SourceData(sourceDataName(target, data.names(), namespace), Map.of(ERROR_PROPERTY, "true"));
 		}
 
-		String names = data.names().stream().sorted().collect(Collectors.joining(PROPERTY_SOURCE_NAME_SEPARATOR));
-		return new SourceData(ConfigUtils.sourceName(target, names, namespace), data.data());
+	}
 
+	/*
+	 * When there is no data, the name of the property source is made from provided
+	 * labels, unlike when the data is present: when we use secret names.
+	 */
+	private SourceData emptySourceData(Map<String, String> labels, String target, String namespace) {
```

Comment:
Wouldn't the secret still have a name in this case?

Reply:
But did the name of the empty property source also have the same problem as non-empty property sources?